modules/gapi/include/opencv2/gapi/own/assert.hpp

```diff
@@ -2,16 +2,23 @@
 // It is subject to the license terms in the LICENSE file found in the top-level directory
 // of this distribution and at http://opencv.org/license.html.
 //
-// Copyright (C) 2018 Intel Corporation
+// Copyright (C) 2018-2020 Intel Corporation
 
 
 #ifndef OPENCV_GAPI_OWN_ASSERT_HPP
 #define OPENCV_GAPI_OWN_ASSERT_HPP
 
+#include <opencv2/gapi/util/compiler_hints.hpp>
+
 #if !defined(GAPI_STANDALONE)
 #include <opencv2/core/base.hpp>
 #define GAPI_Assert CV_Assert
-#define GAPI_DbgAssert CV_DbgAssert
+
+#if defined _DEBUG || defined CV_STATIC_ANALYSIS
+#  define GAPI_DbgAssert CV_DbgAssert
+#else
+#  define GAPI_DbgAssert(expr) cv::util::suppress_unused_warning(expr)
+#endif
```

Comment:
So what's the reason for all these changes?

Reply:
the reason is - compiler warnings on release builds on unused variables in asserts

Reply:
So, this unused variable is still going to be evaluated. Why we need that?

Reply:
It seems to me that actually evaluate the expression is the only reliable way to actually "use"  variables it made of (and there fore silence those warnings). 

And yes compiling optimization of dead code removal is crucial here

Reply:
> "unused"

problem is in caller code. Not in assertion macro.

Reply:
Unfortunately, no the problem is _in_ the macro. Here typical use case  : 
```c++
    struct root_decrement_ref_count{
        void operator()(tbb::task* t) const {
            auto result = t->decrement_ref_count();
            ASSERT(result >= 1);
        }
    };
```  
This functor is used as RAII deleter for `unique::ptr`. Main Invariant of the code is that reference count of the TBB root task is not lesser that 1. This invariant is checked here. Reference count is atomic and the only safe way to get it's previous value save result of the `decrement_ref_count()` call (it Read-Modify-Write operation over atomic).  The problem is that it is used only in the assertion and in the release mode it has to silenced explicitly, which is _ugly_ IMHO.

Reply:
Just replace to C-like `assert()` and you will receive the same "unused" warning in release builds.

Workaround for such cases:
```.diff
-auto result = t->decrement_ref_count();
+auto result = t->decrement_ref_count(); (void)result;
 ASSERT(result >= 1);
```

Reply:
Yes, almost every problem can be workarounded :)  ... but why if there is a proper solution to it ?

Reply:
> if there is a proper solution

It still evaluates expression.
More complex and common cases looks like `ASSERT(node->isValid());`. Compiler's DCE can't remove that completely.

---

Assertion macro must not evaluate passed expression (should be empty) in case of "noop".
Do not confuse developers with new specific behavior.

Reply:
Last version of the patch does not evaluate the expression (and not cause "unused" warning)

Reply:
OK, it is a tricky way, but looks like it works.